render/process.go

```diff
@@ -89,56 +81,67 @@ var ProcessNameRenderer = ConditionalRenderer(renderProcesses,
 	),
 )
 
-// MapEndpoint2Pseudo makes internet of host pesudo nodes from a endpoint node.
-func MapEndpoint2Pseudo(n report.Node, local report.Networks) report.Nodes {
-	_, addr, _, ok := report.ParseEndpointNodeID(n.ID)
-	if !ok {
-		return report.Nodes{}
-	}
-
-	if externalNode, ok := NewDerivedExternalNode(n, addr, local); ok {
-		return report.Nodes{externalNode.ID: externalNode}
-	}
-
-	// due to https://github.com/weaveworks/scope/issues/1323 we are dropping
-	// all non-external pseudo nodes for now.
-	return report.Nodes{}
+// endpoints2Processes joins the endpoint topology to the process
+// topology, matching on hostID and pid.
+type endpoints2Processes struct {
 }
 
-// MapEndpoint2Process maps endpoint Nodes to process
-// Nodes.
-//
-// If this function is given a pseudo node, then it will just return it;
-// Pseudo nodes will never have pids in them, and therefore will never
-// be able to be turned into a Process node.
-//
-// Otherwise, this function will produce a node with the correct ID
-// format for a process, but without any Major or Minor labels.
-// It does not have enough info to do that, and the resulting graph
-// must be merged with a process graph to get that info.
-func MapEndpoint2Process(n report.Node, local report.Networks) report.Nodes {
-	// Nodes without a hostid are treated as pseudo nodes
-	if _, ok := n.Latest.Lookup(report.HostNodeID); !ok {
-		return MapEndpoint2Pseudo(n, local)
-	}
-
-	pid, timestamp, ok := n.Latest.LookupEntry(process.PID)
-	if !ok {
+func (e endpoints2Processes) Render(rpt report.Report, dct Decorator) report.Nodes {
+	if len(rpt.Process.Nodes) == 0 {
```

Comment:
This is rather at odds with the "we have a pid, but no matching process node; create a new one rather than dropping the data" behaviour below, and differs from the existing behaviour.

Reply:
I believe it does match the behaviour of `ProcessRenderer = Memoise(ConditionalRenderer(renderProcesses, ...` - that condition is `len(rpt.Process.Nodes) >= 1`

Reply:
Hmm. Good point. But odd nonetheless.